pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go

```diff
@@ -403,26 +403,86 @@ func (m panicNodeMutator) Generate(
 	rng *rand.Rand, plan *TestPlan, planner *testPlanner,
 ) []mutation {
 	var mutations []mutation
-	maxPanics := len(plan.upgrades)
-	numPanics := 1 + rng.Intn(maxPanics)
-
+	upgrades := randomUpgrades(rng, plan)
 	idx := newStepIndex(plan)
-	possiblePointsInTime := plan.
-		newStepSelector().
-		// We don't want to panic concurrently with other steps, and inserting before a concurrent step
-		// causes the step to run concurrently with that step, so we filter out any concurrent steps.
-		Filter(func(s *singleStep) bool {
-			return s.context.System.Stage >= InitUpgradeStage && !idx.IsConcurrent(s)
-		})
+	nodeList := planner.currentContext.System.Descriptor.Nodes
+
+	for _, upgrade := range upgrades {
+		possiblePointsInTime := upgrade.
+			// We don't want to panic concurrently with other steps, and inserting before a concurrent step
+			// causes the step to run concurrently with that step, so we filter out any concurrent steps.
+			Filter(func(s *singleStep) bool {
+				return s.context.System.Stage >= InitUpgradeStage && !idx.IsConcurrent(s)
+			})
 
-	for range numPanics {
-		nodeList := planner.currentContext.System.Descriptor.Nodes
 		targetNode := nodeList.SeededRandNode(rng)
-		addRandomly := possiblePointsInTime.
-			RandomStep(rng).
-			InsertBefore(panicNodeStep{planner.currentContext.System.Descriptor.Nodes[0], targetNode, planner.rt})
+		stepToPanic := possiblePointsInTime.RandomStep(rng)
+		hasInvalidConcurrentStep := false
+		concurrentStepMarked := false
+		var concurrentStep *singleStep
+
+		isIncompatibleStep := func(s *singleStep) bool {
+			// Restarting the system on a different node while our panicked node is still dead can
+			// cause the cluster to lose quorum, and we cannot run restartWithNewBinaryStep on the
+			// panicked node as it is already dead, so we avoid any system restarts.
+			_, restart := s.impl.(restartWithNewBinaryStep)
+			// Waiting for stable cluster version targets every node in
+			// the cluster, so a node cannot be dead during this step.
+			_, waitForStable := s.impl.(waitForStableClusterVersionStep)
+			// Many hook steps do not support running with a dead node,
+			// so we avoid inserting after a hook step.
+			_, runHook := s.impl.(runHookStep)
+
+			if idx.IsConcurrent(s) {
+				if !concurrentStepMarked {
```

Comment:
nit: maybe  `concurrentStep` -> `firstStepInConcurrentBlock` to clarify what we are using it for? you could also get rid of `concurrentStepMarked` if you did:

```
			if idx.IsConcurrent(s) {
				if firstStepInConcurrentBlock == nil {
					firstStepInConcurrentBlock = s
				}
				hasInvalidConcurrentStep = true
			} else {
				hasInvalidConcurrentStep = false
        firstStepInConcurrentBlock = nil
			}
```

Up to you if you agree that having to deal with one less bool makes it cleaner though.